src/Model/GroundWaterFlow/gwf3vsc8.f90

```diff
@@ -0,0 +1,1519 @@
+                              ctemp, ivisc, a2, a3, a4, auxvar) result(viscbnd)
+! ******************************************************************************
+! get_bnd_viscosity -- Return the viscosity of the boundary package using one of
+!   several different options in the following order of priority:
+!     1. Assign as aux variable in column with name 'VISCOSITY'
+!     2. Calculate using viscosity equation and nviscspecies aux columns
+!     3. If neither of those, then assign as viscref
+! ******************************************************************************
+!
+!    SPECIFICATIONS:
+! ------------------------------------------------------------------------------
+    ! -- modules
+    ! -- dummy
+    integer(I4B), intent(in) :: n
+    integer(I4B), intent(in) :: locvisc
+    real(DP), intent(in) :: a2, a3, a4
+    integer(I4B), dimension(:), intent(in) :: ivisc
+    integer(I4B), dimension(:), intent(in) :: locconc
+    real(DP), intent(in) :: viscref
+    real(DP), dimension(:), intent(in) :: dviscdc
+    real(DP), dimension(:), intent(in) :: cviscref
+    real(DP), dimension(:), intent(inout) :: ctemp
+    real(DP), dimension(:, :), intent(in) :: auxvar
+    ! -- return
+    real(DP) :: viscbnd
+    ! -- local
+    integer(I4B) :: i
+! ------------------------------------------------------------------------------
+    !
+    ! -- assign boundary viscosity based on one of three options
+    if (locvisc > 0) then
+      ! -- assign viscosity to an aux column named 'VISCOSITY'
+      viscbnd = auxvar(locvisc, n)
+    else if (locconc(1) > 0) then
+      ! -- calculate viscosity using one or more concentration auxcolumns
+      do i = 1, size(locconc)
+        ctemp(i) = DZERO
+        if (locconc(i) > 0) then
+          ctemp(i) = auxvar(locconc(i), n)
+        end if
+      end do
+      viscbnd = calc_visc(ivisc, viscref, dviscdc, cviscref, ctemp, a2, a3, a4)
+    else
+      ! -- neither of the above, so assign as viscref
+      viscbnd = viscref
+    end if
+    !
+    ! -- return
+    return
+  end function calc_bnd_viscosity
+
+  !> @brief Account for viscosity in horizontal flow barriers
+  !!
+  !! Will return the viscosity associated with the upgradient node (cell)
+  !! to the HFB package for adjusting the hydraulic characteristic (hydchr)
+  !! of the barrier
+  !<
+  subroutine set_hfb_visc(this, cellid, hfbviscratio)
+    ! -- dummy variables
+    class(GwfVscType) :: this
+    integer(I4B), intent(in) :: cellid
+    ! -- return
+    real(DP), intent(inout) :: hfbviscratio
+    ! -- local
+    real(DP) :: visc
+! ------------------------------------------------------------------------------
+    !
+    ! -- Retrieve viscosity for the passed node number
+    visc = this%visc(cellid)
+    !
+    ! -- Calculate the viscosity ratio for the
+    hfbviscratio = calc_vsc_ratio(this%viscref, visc)
+    !
+    ! -- return
+    return
+  end subroutine set_hfb_visc
```

Comment:
Agreed.  Renamed to `calc_q_visc` (shorthand for calculate flow viscosity).  Since HFB is a flow package and not a boundary package (i.e., doesn't inherit from `BndType`, the only to packages that use it are NPF and HFB.  Code was altered in `update_k_with_vsc` to also make use of `calc_q_visc` to remove a redundant calculation.  Change will be reflected in next commit